openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+template<typename GridT>
+typename GridT::Ptr
+fogToSdf(const GridT &fogGrid,
+         typename GridT::ValueType isoValue,
+         int nIter = 1);
+
+/// @brief Given an existing approximate SDF it solves the Eikonal equation for all its
+///        active voxels. Active input voxels with a signed distance value above the
+///        given isoValue will have POSITIVE distance values on output, i.e. they are
+///        assumed to be OUTSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input sdf volume.
+///
+/// @param sdfGrid  An approximate signed distance field to the specified iso-surface.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a sdfGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note The only difference between this method and fogToSdf, defined above, is the
+///       convention of the sign of the output distance field.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
+///         voxels or if it intersects any active tiles in @a sdfGrid.
+template<typename GridT>
+typename GridT::Ptr
+sdfToSdf(const GridT &sdfGrid,
+         typename GridT::ValueType isoValue = 0,
+         int nIter = 1);
+
+/// @brief Computes the extension of a scalar field, defined by the specified functor,
+///        off an iso-surface from an input FOG volume.
+///
+/// @return A shared pointer to the extension field defined from the active values in
+///         the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param op       Functor with signature [](const Vec3R &xyz)->float that
+///                 defines the Dirichlet boundary condition, on the iso-surface,
+///                 of the field to be extended.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note Strictly speaking a fog volume is normalized to the range [0,1] but this
+///       method accepts a scalar volume with an arbitary range, as long as the it
+///       includes the @a isoValue.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
```

Comment:
In this case the grid should be set to the background value and not be a runtime error.

Reply:
what background value? This method takes in a density, and return a scalar field, with values defined by the functor, that is extended from an iso-surface of the density. If not iso-surface exists then there is no meaningful output in my opinion. For instance, what should the background value me? The only alternative to throwing is to return a null shared pointer. Would you prefer that?

Reply:
I'd like for the resturned grid to have the same topology as the fog grid.  That is a nice invariant for the caller to work with.

I think we should add a typename GridT::ValueType backgroundValue to answer exactly the question of what the extension is.   (Or maybe not GridT::ValueType, as that is always scalar, but derived from OpT?)

One thing this doesn't let you specify is the type of grid to fill...  OpT is an opaque functor of unspecified return type, so how can you use this for vector grids?   If we had a parameter that is the background value, that parameter can be the GridT of the generated grid?

Hmm...  I actually don't like that very much as it runs into the 0 problem.

Probably best to prefix an explicit template on the extrap-gridtype.

template<typename ResultT, typename GridT, typename OpT>
￼typename Grid<ResultT>::Ptr
￼fogToExt(const GridT &fogGrid,
￼         const OpT &op,
￼         typename GridT::ValueType isoValue,
         ResultT backgroundValue,
￼         int nIter = 1);

Reply:
see comment above - I'm no longer throwing